### DISCUSSION TITLE
Ping return values are wrong and non linear with congestion

### DISCUSSION THREAD

**Discussion by dmfiore** [openthread/openthread#7792] [Q&A] [answered]
Hi,
I did a quick search, and didn't find anything, was wondering if there a known issue on the ping value returns being wrong ?
I find they are accurate with low values and no congestion, however when under congestion, the values being returned can be 15k to 50k of ms  (ie, 15-50 seconds).   For which this return is being returned with a few seconds, so that is not possible.
For example, let's say I point 10 devices to ping x100 of size 512 bytes with inter packet delay of say 3000ms.   The ping times being printed to console for the 100 iterations can range up to 60000, but it reports this every 3000 ms.
Thanks for any pointers on this,
-D

**Comment by jwhui** [comment, marked as answer]
Can you provide a copy of the console output, including the command you used and the output you are observing? Thanks.

**Reply by dmfiore** [reply]
Thanks @jwhui , I can.  Let me make a run tomorrow and get full console.  
I don't really look through the source code, and test more as black box.   For that, my suspicion is that when you issue a ping command for say 100 iterations with 2sec timeout, the consecutive return times are not being reset when a packet is lost or not within the ping api timeout initial use for times x.

**Reply by dmfiore** [reply]
An example would be this file that issues ping for x100 iterations for size=1024 over 9 devices.  These devices reported on console for each of the 100 iterations.  Given that each ping is sent every 5000 ms, there is no way the return values of max of 50,000 ms makes sense.
I will have to get OT console.
![image](https://user-images.githubusercontent.com/58004888/172255766-b88bb2fb-1054-47ea-8a32-3bd4a10c96f5.png)

**Reply by dmfiore** [reply]
@jwhui Here's a console of x9 devices transmitting ping to a single device.   Way too much traffic as sending 1024 byte packets every second from all 9 devices.   Notice the console returns for the ping every second on each device keeps increasing during congestion from initial ~150ms to >25,000 ms on the end of the 100 iteration tests.   The console is time stamped, but busy because it is the 9 devices console combined on test PC.   Notice that from one device it is reporting each discrete return value being much greater than the 1 sec reporting, ie, reporting with 1 sec that it took 20 secs.

Each 9 devices were issued this command (with a jitter delay of ~150ms between each call:
>ping fdde:ad00:beef:0:8a24:2b89:955:800e 256 100 1.0 3 2
[OT-ping-values-floating-on-congestion.txt](https://github.com/openthread/openthread/files/8855350/OT-ping-values-floating-on-congestion.txt)

**Reply by jwhui** [reply]
I have reviewed the log. I believe that because the offered load is more than the network can handle, the round-trip time increases during the test. So I think what you are observing is expected.

**Reply by dmfiore** [reply]
I'm not sure that is accurate though.
If I send x10 iterations each 1000ms apart and those x10 iterations report each took 25,000ms and the total time would be = 250 seconds for which all of this 250 seconds was reported within 10sec of realtime.

**Reply by dmfiore** [reply]
I was hoping to find thresholds of what the network can handle.  And this impedes that.   But certainly we should use something not based on ICMP.  That said, I'm pretty sure we could change the ping API logic here for the better and perhaps reset the values and be more accurate, else we should perhaps note that ping API is only accurate if there is no congestion ?

**Reply by jwhui** [reply]
ICMPv6 Echo Request messages are sent without waiting to receive replies to prior Echo Request messages, so there is overlap in the delays for each individual request/response transaction.

The round-trip time computed with the ping command is intended to measure the latency of the network with or without congestion. This is common practice with any IP-based network.

Increased latency under load is a common issue due to network buffering/queueing - this is often referred to as [bufferbloat](https://en.wikipedia.org/wiki/Bufferbloat).

**Reply by dmfiore** [reply]
Thanks for the quick responses and info, Jonathan.   I still think we have a bug here.
If you send x10 packets with 1 sec packet interval(ie rate) for a total time of transmitting those commands over 10 secs.  
Then how can each of those 10 replies within that 10 seconds add up to 250 seconds in total ?  It only took 10 seconds for all those commands to be executed and completed with values of time.
What I can't get my head around is that if a ping reports it took 25secs, why didn't the console just wait 25secs for that value versus it being printed out within 1 sec.

**Reply by jwhui** [reply]
The round-trip time is printed whenever an ICMPv6 Echo Reply is received. The Echo Request was sent 25 seconds in the past.

Note that in your log, the first received Echo Reply had an icmp_seq value of 201 and the last received reply had an icmp_seq value of 265. As a result, the log is not showing the full 100 packets that were sent.
